File: 01 Leet - Easy/53-maximum-subarray.js
/*

Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.

If you have figured out the O(n) solution, try coding another solution using the divide and conquer approach, which is more subtle.

*/

let array = [-2, 1, -3, 4, -1, 2, 1, -5, 4];

let maxSubArray = function(nums) {
  let n = nums.length;
  let maxSum = nums[0];
    for(let i = 1; i < n; i++) {
      if (nums[i - 1] > 0) {
        nums[i] += nums[i - 1];
        console.log('after', nums)
      }
      maxSum = Math.max(nums[i], maxSum);
    }
    return maxSum;
};

console.log(maxSubArray(array));

File: 01 - Easy/53-maximum-subarray.py
'''
Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.

Follow up:

If you have figured out the O(n) solution, try coding another solution using the divide and conquer approach, which is more subtle.
'''


class Solution:
    def maxSubArray(self, nums):
        maxSum = nums[0]
        for i in range(1, len(nums)):
            # Only sum if positive, this way we gather the biggest sum
            if nums[i - 1] > 0:
                nums[i] += nums[i - 1]
            # print('array:', nums)
            maxSum = max(nums[i], maxSum)
        return maxSum


input = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
sol = Solution()
print(sol.maxSubArray(input))
